// eleventy-plugin-china-mirror/index.js
/**
 * Eleventy Plugin China Mirror
 * Zero-config China network acceleration for Eleventy
 * 
 * Features:
 * - Auto-configures npm registry to Chinese mirror
 * - Intercepts fetch requests to cdn.11ty.dev and routes through mirror
 * - Zero configuration required
 */

import { existsSync, writeFileSync } from 'node:fs';
import { join } from 'node:path';
import { fileURLToPath } from 'node:url';

// Plugin configuration
const CONFIG = {
  npmrcContent: `# Eleventy China Mirror - Auto-generated configuration
registry=https://registry.npmmirror.com
sharp_binary_host=https://npmmirror.com/mirrors/sharp
sharp_libvips_binary_host=https://npmmirror.com/mirrors/sharp-libvips
disturl=https://npmmirror.com/mirrors/node
`,
  originalCdn: 'cdn.11ty.dev',
  mirrorCdn: 'registry.npmmirror.com/11ty-cdn',
  logPrefix: '[china-mirror]'
};

/**
 * Check if .npmrc exists and create it if not
 */
function setupNpmrc(projectRoot) {
  const npmrcPath = join(projectRoot, '.npmrc');
  
  if (!existsSync(npmrcPath)) {
    try {
      writeFileSync(npmrcPath, CONFIG.npmrcContent, 'utf8');
      console.log(`${CONFIG.logPrefix} .npmrc created successfully - npm registry configured to Chinese mirror`);
      return true;
    } catch (error) {
      console.warn(`${CONFIG.logPrefix} Failed to create .npmrc:`, error.message);
      return false;
    }
  } else {
    console.log(`${CONFIG.logPrefix} .npmrc already exists - skipping configuration`);
    return false;
  }
}

/**
 * Intercept and modify fetch requests to use Chinese mirror
 */
function setupFetchInterceptor() {
  const originalFetch = globalThis.fetch;
  
  if (originalFetch && !originalFetch.__chinaMirrorPatched) {
    globalThis.fetch = async function(url, ...opts) {
      let modifiedUrl = url;
      
      // Convert URL to string if it's a URL object
      const urlString = typeof url === 'string' ? url : url.toString();
      
      // Replace cdn.11ty.dev with Chinese mirror
      if (urlString.includes(CONFIG.originalCdn)) {
        modifiedUrl = urlString.replace(CONFIG.originalCdn, CONFIG.mirrorCdn);
        console.log(`${CONFIG.logPrefix} Fetch intercepted: ${urlString} → ${modifiedUrl}`);
      }
      
      try {
        // Try the modified URL first
        const res = await originalFetch(modifiedUrl, ...opts);
        // Auto-fallback on non-2xx responses when using mirror
        if (!res.ok && modifiedUrl !== urlString) {
          console.warn(`${CONFIG.logPrefix} Mirror responded with ${res.status}, falling back to original: ${urlString}`);
          return await originalFetch(url, ...opts);
        }
        return res;
      } catch (error) {
        // If mirror fails, fallback to original URL
        if (modifiedUrl !== urlString) {
          console.warn(`${CONFIG.logPrefix} Mirror failed, falling back to original:`, error.message);
          return await originalFetch(url, ...opts);
        }
        throw error;
      }
    };
    
    globalThis.fetch.__chinaMirrorPatched = true;
    console.log(`${CONFIG.logPrefix} Fetch interceptor installed - cdn.11ty.dev requests will be routed through mirror`);
  }
}

/**
 * Main plugin function for Eleventy
 */
function eleventyPluginChinaMirror(eleventyConfig, options = {}) {
  // Determine project root (current working directory)
  const projectRoot = process.cwd();
  
  console.log(`${CONFIG.logPrefix} Initializing China Mirror plugin...`);
  
  // Setup npmrc configuration
  setupNpmrc(projectRoot);
  
  // Setup fetch interceptor
  setupFetchInterceptor();
  
  // Log successful initialization
  console.log(`${CONFIG.logPrefix} Plugin initialized successfully!`);
  console.log(`${CONFIG.logPrefix} Benefits:`);
  console.log(`${CONFIG.logPrefix}   - npm dependencies will download from Chinese mirrors`);
  console.log(`${CONFIG.logPrefix}   - cdn.11ty.dev requests will be accelerated`);
  console.log(`${CONFIG.logPrefix}   - No manual configuration required`);
}

// Export the plugin function for Eleventy
export default eleventyPluginChinaMirror;

// Also export individual functions for testing
export { setupNpmrc, setupFetchInterceptor, CONFIG };
