// eleventy-plugin-china-mirror/index.cjs
const { existsSync, writeFileSync } = require('node:fs');
const { join } = require('node:path');

const CONFIG = {
  npmrcContent: `# Eleventy China Mirror - Auto-generated configuration
registry=https://registry.npmmirror.com
sharp_binary_host=https://npmmirror.com/mirrors/sharp
sharp_libvips_binary_host=https://npmmirror.com/mirrors/sharp-libvips
disturl=https://npmmirror.com/mirrors/node
`,
  originalCdn: 'cdn.11ty.dev',
  mirrorCdn: 'registry.npmmirror.com/11ty-cdn',
  logPrefix: '[china-mirror]'
};

function setupNpmrc(projectRoot) {
  const npmrcPath = join(projectRoot, '.npmrc');
  if (!existsSync(npmrcPath)) {
    try {
      writeFileSync(npmrcPath, CONFIG.npmrcContent, 'utf8');
      console.log(`${CONFIG.logPrefix} .npmrc created successfully - npm registry configured to Chinese mirror`);
      return true;
    } catch (error) {
      console.warn(`${CONFIG.logPrefix} Failed to create .npmrc:`, error.message);
      return false;
    }
  } else {
    console.log(`${CONFIG.logPrefix} .npmrc already exists - skipping configuration`);
    return false;
  }
}

function setupFetchInterceptor() {
  const originalFetch = globalThis.fetch;
  if (originalFetch && !originalFetch.__chinaMirrorPatched) {
    globalThis.fetch = async function(url, ...opts) {
      let modifiedUrl = url;
      const urlString = typeof url === 'string' ? url : url.toString();
      if (urlString.includes(CONFIG.originalCdn)) {
        modifiedUrl = urlString.replace(CONFIG.originalCdn, CONFIG.mirrorCdn);
        console.log(`${CONFIG.logPrefix} Fetch intercepted: ${urlString} → ${modifiedUrl}`);
      }
      try {
        const res = await originalFetch(modifiedUrl, ...opts);
        if (!res.ok && modifiedUrl !== urlString) {
          console.warn(`${CONFIG.logPrefix} Mirror responded with ${res.status}, falling back to original: ${urlString}`);
          return await originalFetch(url, ...opts);
        }
        return res;
      } catch (error) {
        if (modifiedUrl !== urlString) {
          console.warn(`${CONFIG.logPrefix} Mirror failed, falling back to original:`, error.message);
          return await originalFetch(url, ...opts);
        }
        throw error;
      }
    };
    globalThis.fetch.__chinaMirrorPatched = true;
    console.log(`${CONFIG.logPrefix} Fetch interceptor installed - cdn.11ty.dev requests will be routed through mirror`);
  }
}

function eleventyPluginChinaMirror(eleventyConfig, options = {}) {
  const projectRoot = process.cwd();
  console.log(`${CONFIG.logPrefix} Initializing China Mirror plugin...`);
  setupNpmrc(projectRoot);
  setupFetchInterceptor();
  console.log(`${CONFIG.logPrefix} Plugin initialized successfully!`);
  console.log(`${CONFIG.logPrefix}   - npm dependencies will download from Chinese mirrors`);
  console.log(`${CONFIG.logPrefix}   - cdn.11ty.dev requests will be accelerated`);
  console.log(`${CONFIG.logPrefix}   - No manual configuration required`);
}

module.exports = eleventyPluginChinaMirror;
module.exports.setupNpmrc = setupNpmrc;
module.exports.setupFetchInterceptor = setupFetchInterceptor;
module.exports.CONFIG = CONFIG;
